src/resources/License.py

```diff
@@ -0,0 +1,328 @@
+                    'gasPrice': self.web3.to_wei('100', 'gwei')
+                }
+            )
+            logger.info(f"Built transaction: {transaction}")
+
+            # Sign the transaction using the account object
+            signed_txn = self.account.sign_transaction(transaction)
+            logger.info(f"Signed transaction: {signed_txn}")
+
+            # Send the transaction
+            tx_hash = self.web3.eth.send_raw_transaction(signed_txn.rawTransaction)
+            logger.info(f"Transaction hash: {tx_hash.hex()}")
+
+            # Wait for transaction receipt with a longer timeout
+            tx_receipt = self.web3.eth.wait_for_transaction_receipt(tx_hash, timeout=600)
+            logger.info(f"Transaction receipt: {tx_receipt}")
+
+            # Parse the event logs for LicenseTermsRegistered
+            if not tx_receipt.logs:
+                logger.error(f"No logs found in transaction receipt: {tx_receipt}")
+                return None
+
+            target_logs = self._parse_tx_license_terms_registered_event(tx_receipt)
+            return {
+                'txHash': tx_hash.hex(),
+                'licenseTermsId': target_logs
+            }
+
+        except Exception as e:
+            logger.error(f"Error interacting with contract: {e}")
+            raise e
+
+    def _parse_tx_license_terms_registered_event(self, tx_receipt):
+        event_signature = self.web3.keccak(text="LicenseTermsRegistered(uint256,address,bytes)").hex()
+
+        for log in tx_receipt['logs']:
+            if log['topics'][0].hex() == event_signature:
+                return int(log['topics'][1].hex(), 16)
+
+        return None
+    
+    def attachLicenseTerms(self, ip_id, license_template, license_terms_id):
+        try:
+            logger.info(f"Starting attachLicenseTerms with ip_id: {ip_id}, license_template: {license_template}, license_terms_id: {license_terms_id}")
+
+            # Validate the license template address
+            if not Web3.is_address(license_template):
+                raise ValueError(f'Address "{license_template}" is invalid.')
+
+            # Check if the IP is registered
+            is_registered = self.ip_asset_registry_client.isRegistered(ip_id)
+            if not is_registered:
+                raise ValueError(f"The IP with id {ip_id} is not registered.")
+
+            # Check if the license terms exist
+            is_existed = self.license_registry_client.exists(license_template, license_terms_id)
+            if not is_existed:
+                raise ValueError(f"License terms id {license_terms_id} do not exist.")
+
+            # Check if the license terms are already attached to the IP
+            is_attached_license_terms = self.license_registry_client.hasIpAttachedLicenseTerms(ip_id, license_template, license_terms_id)
+            if is_attached_license_terms:
+                raise ValueError(f"License terms id {license_terms_id} is already attached to the IP with id {ip_id}.")
```

Comment:
ideally we don't raise error here. Check how the typescript SDK does it

Reply:
typescript sdk raises error if already attached